test/hotspot/jtreg/compiler/lib/verify/Verify.java

```diff
@@ -25,33 +25,80 @@
 
 import java.util.Optional;
 import java.lang.foreign.*;
+import java.lang.reflect.Method;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+
 
 /**
- * The {@link Verify} class provides a single {@link Verify#checkEQ} static method, which recursively
- * compares the two {@link Object}s by value. It deconstructs {@link Object[]}, compares boxed primitive
- * types, and compares the content of arrays and {@link MemorySegment}s.
+ * The {@link Verify} class provide {@link Verify#checkEQ} and {@link Verify#checkEQWithRawBits},
+ * which recursively compare the two  {@link Object}s by value. They deconstruct an array of objects,
+ * compare boxed primitive types, compare the content of arrays and {@link MemorySegment}s, and check
+ * that the messages of two {@link Exception}s are equal. They also checks for the equivalent content
+ * in {@code Vector}s from the Vector API.
+ *
+ * <p>
+ * Further, they compare Objects from arbitrary classes, using reflection. We check the fields of the
+ * Objects, and compare their recursive structure. Since we use reflection, this can be slow.
+ *
+ * <p>
+ * When a comparison fails, then methods print helpful messages, before throwing a {@link VerifyException}.
  *
- * When a comparison fail, then methods print helpful messages, before throwing a {@link VerifyException}.
+ * <p>
+ * We have to take special care of {@link Float}s and {@link Double}s, since they both have various
+ * encodings for NaN values while the Java specification regards them as equal. Hence, we
+ * have two modes of comparison. With {@link Verify#checkEQ} different NaN values are regarded as equal.
+ * This applies to the boxed floating types, as well as arrays of floating arrays. With
+ * {@link Verify#checkEQWithRawBits} we compare the raw bits, and so different NaN encodings are not equal.
+ * Note: {@link MemorySegment} data is always compared with raw bits.
  */
 public final class Verify {
+    private final boolean isFloatCheckWithRawBits;
 
-    private Verify() {}
+    /**
+     * When comparing arbitrary classes recursively, we need to remember which
+     * pairs of objects {@code (a, b)} we have already visited. The maps
+     * {@link a2b} and {@link b2a} track these edges. Caching which pairs
```

Comment:
I think it's fine to use `code` here since the Javadocs links to itself otherwise.
```suggestion
     * {@code a2b} and {@code b2a} track these edges. Caching which pairs
```